File: 2021/12.rb
INPUT = File.read(__FILE__.sub('.rb', '.txt')).lines.map(&:chomp)

ADJACENTS = Hash.new { |h, k| h[k] = [] }

INPUT.map { |line| line.split("-") }.each do |a, b|
  ADJACENTS[a] << b
  ADJACENTS[b] << a
end

def find_all_paths(make_an_exception)
  stack = []
  stack.push(['start', ['start'], {'start' => true}, false])

  res = []

  until stack.empty?
    node = stack.pop
    cave, path, visited, exception = node

    count_caves = path.group_by(&:itself)
    lowercase_caves = path.select { |x| x.downcase == x && x != 'start' && x != 'end' }

    if cave == 'end'
      res << path
      next
    end

    ADJACENTS[cave].each do |next_cave|
      next_visited = visited.merge(next_cave => next_cave.downcase == next_cave)
      next_path = path + [next_cave]
      can_make_an_exception = make_an_exception && !exception && next_cave != 'start' && next_cave != 'end'

      if !visited[next_cave]
        stack.push([next_cave, next_path, next_visited, exception])
      elsif can_make_an_exception
        stack.push([next_cave, next_path, next_visited, true])
      end
    end
  end

  res
end

def part1
  find_all_paths(false)
end

def part2
  find_all_paths(true)
end

puts part1.length
puts part2.length

File: 2021/12.js
const fs = require("fs")

const INPUT = fs.readFileSync("12.txt", "utf8").split("\n").filter(line => line)
const ADJACENTS = {}

INPUT.forEach(line => {
  const [a, b] = line.split("-")

  ADJACENTS[a] ||= []
  ADJACENTS[b] ||= []
  ADJACENTS[a].push(b)
  ADJACENTS[b].push(a)
})

const CACHE = {}

const findAllPaths = function(start, path, visited, makeAnException) {
  const cacheKey = [start, Object.keys(visited).filter(k => visited[k]).sort(), makeAnException].join(";")

  let res = CACHE[cacheKey]
  if(res === undefined) res = findAllPathsHelper(start, path, visited, makeAnException)

  CACHE[cacheKey] = res

  return res
}

const findAllPathsHelper = function(start, path, visited, makeAnException) {
  if(start === "end") return [path.map(node => node)]

  let res = []

  ADJACENTS[start].forEach(adjacent => {
    const nextMakeAnException = makeAnException && !visited[adjacent]
    const canMakeAnException = makeAnException && (adjacent !== "start") && (adjacent !== "end")

    if(!visited[adjacent] || canMakeAnException) {
      const wasVisited = visited[adjacent]
      if(adjacent.toLowerCase() === adjacent) visited[adjacent] = true
      path.push(adjacent)

      res = res.concat(findAllPaths(adjacent, path, visited, nextMakeAnException))

      path.pop()
      visited[adjacent] = wasVisited
    }
  })

  return res
}

console.log(findAllPaths("start", ["start"], { start: true }, false).length)
console.log(findAllPaths("start", ["start"], { start: true }, true).length)
